---
title: "edX Intro to Bioconductor section 3: Genomic annotation with Bioconductor"
author: Josh Cook
date: 2018-06-06
output: html_notebook
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE, results='hide'}
library(knitr)
library(Biobase)
```

# Prologue: Detailed applications of GRanges

## Introduction to GenomicRanges

We will be using a library made for this course: `ERBS`.

```{r, warning=FALSE, message=FALSE}
library(ERBS)
# estrogen binding site data from two cell lines
data("GM12878")
data("HepG2")
```

We can see `HepG2` is a `GRanges` object:

```{r}
HepG2
```

Get the data as a dataframe using the function `values`.  

Access data by treating the object as a matrix. For example get the first 10 rows using `HepG2[1:10, ]`.  

Get chromosomes using the function `seqnames(HepG2)`. This returns an `rle` that can be use like a normal `list` object. The `rle` class is a good storage device, especially if the sequence names were sorted.  

Can order the object by chromosome and then location using the command `HepG2[order(HepG2),]`.

## Introduction to Genomic Ranges Assessment

1. In the video we used the `values` method to extract meta-data on the regions. An alternative, and  preferred approach, is `mcols`. What is the median of the `signalValue` column for the HepG2 data?

```{r}
df <- mcols(HepG2)
median(df$signalValue)
```

2. In what chromosome is the region with the highest `signalValue` (copy and paste your answer)?

```{r}
HepG2[which(df$signalValue == max(df$signalValue)), ]
```

3. How many regions are from chromosome 16?

```{r}
sum(seqnames(HepG2) == 'chr16')
```

Make a histogram of the widths of the regions from all chromosomes (not just chr16). Note it has a heavy right tail. What is the median width?

```{r}
hist(width(HepG2))
```

```{r}
median(width(HepG2))
```

## Genes as GRanges

We will be using the human genomic data:

```{r, message=FALSE, warning=FALSE}
library(Homo.sapiens)
```

Can get all of the genes as GRanges:

```{r, cache=TRUE}
ghs <- genes(Homo.sapiens)
ghs
```

A usefull function `precede`: "For each range in x, precede returns the index of the range in subject that is directly preceded by the range in x. Overlapping ranges are excluded. NA is returned when there are no qualifying ranges in subject."

```{r}
res <- findOverlaps(HepG2, GM12878)
index <- queryHits(res)
erbs <- HepG2[index, ]
erbs <- GRanges(erbs)

index <- precede(erbs, ghs)
ghs[index[1:3]]
erbs[1:3]
```

## Genes as GRanges Assessment

1. How many genes are represented in `ghs`?

```{r}
length(seqnames(ghs))
```

2. What is the chromosome with the most genes?

```{r}
sort(table(seqnames(ghs)), decreasing=T)[1:5]
```

3. Make a histogram of the widths of genes (use the width() on the GRanges object). This width gives the number of basepairs from the start of the gene to the end, so including exons and introns. Which best describes the width of genes?

```{r}
hist(width(ghs))
```

4. What is the median gene width?

```{r}
median(width(ghs))
```

## Finding the Nearest Gene