---
title: "edX Intro to Bioconductor section 3: Genomic annotation with Bioconductor"
author: Josh Cook
date: 2018-06-06
output: html_notebook
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE, results='hide'}
library(knitr)
library(Biobase)
```

# Prologue: Detailed applications of GRanges

## Introduction to GenomicRanges

We will be using a library made for this course: `ERBS`.

```{r, warning=FALSE, message=FALSE}
library(ERBS)
# estrogen binding site data from two cell lines
data("GM12878")
data("HepG2")
```

We can see `HepG2` is a `GRanges` object:

```{r}
HepG2
```

Get the data as a dataframe using the function `values`.  

Access data by treating the object as a matrix. For example get the first 10 rows using `HepG2[1:10, ]`.  

Get chromosomes using the function `seqnames(HepG2)`. This returns an `rle` that can be use like a normal `list` object. The `rle` class is a good storage device, especially if the sequence names were sorted.  

Can order the object by chromosome and then location using the command `HepG2[order(HepG2),]`.

## Introduction to Genomic Ranges Assessment

1. In the video we used the `values` method to extract meta-data on the regions. An alternative, and  preferred approach, is `mcols`. What is the median of the `signalValue` column for the HepG2 data?

```{r}
df <- mcols(HepG2)
median(df$signalValue)
```

2. In what chromosome is the region with the highest `signalValue` (copy and paste your answer)?

```{r}
HepG2[which(df$signalValue == max(df$signalValue)), ]
```

3. How many regions are from chromosome 16?

```{r}
sum(seqnames(HepG2) == 'chr16')
```

Make a histogram of the widths of the regions from all chromosomes (not just chr16). Note it has a heavy right tail. What is the median width?

```{r}
hist(width(HepG2))
```

```{r}
median(width(HepG2))
```

## Genes as GRanges

We will be using the human genomic data:

```{r, message=FALSE, warning=FALSE}
library(Homo.sapiens)
```

Can get all of the genes as GRanges:

```{r, cache=TRUE}
ghs <- genes(Homo.sapiens)
ghs
```

A usefull function `precede`: "For each range in x, precede returns the index of the range in subject that is directly preceded by the range in x. Overlapping ranges are excluded. NA is returned when there are no qualifying ranges in subject."

```{r}
res <- findOverlaps(HepG2, GM12878)
index <- queryHits(res)
erbs <- HepG2[index, ]
erbs <- GRanges(erbs)

index <- precede(erbs, ghs)
ghs[index[1:3]]
erbs[1:3]
```

## Genes as GRanges Assessment

1. How many genes are represented in `ghs`?

```{r}
length(seqnames(ghs))
```

2. What is the chromosome with the most genes?

```{r}
sort(table(seqnames(ghs)), decreasing=T)[1:5]
```

3. Make a histogram of the widths of genes (use the width() on the GRanges object). This width gives the number of basepairs from the start of the gene to the end, so including exons and introns. Which best describes the width of genes?

```{r}
hist(width(ghs))
```

4. What is the median gene width?

```{r}
median(width(ghs))
```

## Finding the Nearest Gene

Find the gene nearest to each binding site:

```{r}
dis <- distance(erbs, ghs[index])
```

Can define distance differently by finding the nearest transcription start site. USe the `resize` ghs to one basepair at the beginning (ie. the TSS). The `distanceToNearest` to find the distance for each range in `x` to its nearest neighbor in  `subject`.

```{r}
tss <- resize(ghs, 1)
distss <- distanceToNearest(x=erbs, subject=tss)
queryHits(distss)
```

```{r}
dists <- values(distss)$distance
```

Use a histogram to see the distances:

```{r}
hist(dists, nc=100)
```

Can get the genes that are close the binding sites:

```{r}
index <- subjectHits(distss)[dists < 1000]
```

To find out what keys are available in the `Homo.sapiens` db, can use the function `keytpes(Homo.sapiens)`. `GENEID` is one of the options, which is also in the `tss` object. Also use the `columns(Homo.sapiens)` command to see the available columns we can return.

Use the `select` function to query the Homo sapiens db:

```{r}
keys <- as.character(values(tss[index])$GENEID)
select(Homo.sapiens, keys=keys, keytype="GENEID", columns=c("SYMBOL", "GENENAME"))
```

## Finding and getting annotation for closest gene Assessment

In this assessment (which deals with topics in several videos) we will find the closest genes to some of our binding sites. We will use a consensus set of regions. In the video we did it like this:

```{r}
library(Homo.sapiens)
ghs = genes(Homo.sapiens)

res = findOverlaps(HepG2,GM12878)
erbs = HepG2[queryHits(res)]
erbs = granges(erbs)
erbs2= intersect(HepG2,GM12878)
```

2. Using what you learned in the video, convert the `ghs` object to one that represents just the `tss`.

What is the TSS (Transcription Start Site) of the gene with ID: 100113402?

```{r}
tss <- resize(ghs, 1)
tss["100113402",]
```

3. Now using the erbs regions defined in a previous question:

```{r}
res = findOverlaps(HepG2,GM12878)
erbs = HepG2[queryHits(res)]
erbs = granges(erbs)
```

What is the GENEID of the gene with TSS closest to the 4th region of `erbs`?

```{r}
n <- nearest(erbs, tss)
tss[n]
```

4. In the question above, you identified a gene. Use the `select` function to determine which is the `SYMBOL` of this gene.

```{r}
select(Homo.sapiens, keys="2101", keytype="GENEID", columns=c("SYMBOL"))
```

## Getting the Sequence of Regions

In this video, we're going to show how to use the full reference genomic sequence of Homo sapiens in order to look at the content of the binding sites for the estrogen receptor.

```{r, message=FALSE, warning=FALSE}
library(BSgenome.Hsapiens.UCSC.hg19)
Hsapiens
```

Can use the `getSeq` function to extract a set of sequences (or subsequences) from a sequence container like a `BSgenome` object.

```{r}
hepseq <- getSeq(Hsapiens, HepG2)
```

This returns a collection of DNA strings for each binding site in `HepG2`.  

Let's make a control set of strings:

```{r}
rhepseq <- getSeq(Hsapiens, shift(HepG2, 2500))
```

Now we will count the number of appearances of the motif `mot` using `vcountPattern`.

```{r}
mot <- "TCAAGGTC"
sum(vcountPattern(mot, hepseq))
sum(vcountPattern(mot, reverseComplement(hepseq)))
```

```{r}
sum(vcountPattern(mot, rhepseq))
sum(vcountPattern(mot, reverseComplement(rhepseq)))
```

In a real analysis, we would need to account for deviations from the motif using a matrix representation through the package `MotifDb` and a pattern matching programs such as `MEME` or `FIMO` (I'm not sure these are available in R).

## Getting Sequence Assessment

```{r, message=FALSE, warning=FALSE}
library(ERBS)
library(GenomicRanges)
data(HepG2)
data(GM12878)
res = findOverlaps(HepG2,GM12878)
erbs = HepG2[queryHits(res)]
erbs = granges(erbs)
library(BSgenome.Hsapiens.UCSC.hg19)
```

2. Now use the `getSeq` function to extract the sequence of each region in `erbs`. Then compute the GC-content (the number of C's + the number of G's divided by the length of sequence) of each. What is the median GC-content?

```{r}
seqs <- getSeq(Hsapiens,erbs)
gc <- alphabetFrequency(seqs)[,2:3]
n <- width(erbs)
gccontent <- rowSums(gc)/n
median(gccontent)
```

3. Now create a control set of regions by shifting erbs by 10000. What is the median GC-content of these control regions?

```{r}
seqs <- getSeq(Hsapiens, shift(erbs, 10000))
gc <- alphabetFrequency(seqs)[, 2:3]
n <- width(erbs)
gccontent <- rowSums(gc)/n
median(gccontent)
```

# Annotation of genes and transcripts  

## Reference sequence packages for model organisms