---
title: "edX Intro to Bioconductor section 1: Biology background, data examples"
author: Josh Cook
date: 2018-06-06
output: html_notebook
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)
library(BiocInstaller)
library(Biobase)
```

# Section 1: Biology Background, data examples
## Getting Started
### Getting Help
Update bioconductor packages by calling `bioclite()`.
```{r, message=FALSE, warning=FALSE, eval=FALSE}
biocLite()
```
Open vignettes using `vignette()` or `browseVignettes()` (opens in browser as html).

## Introduction to Microarray Technologies
### Microarray Technology 1: How Hybridization Works
Will only measure RNA for which there exists a probe.  
Differences between platforms:
* different densitities of probes (some have space between each pixel, too)
* one or two colors
Main platforms are Affymetrix, Agilent, and Illumina (uses beads instead of chip).  

### Microarray Technology 2: How Microarrays Work
Two-color arrays return a tuple for each pixel (probe) with a pair of values for each color.

### Microarray Technologies 3: Applications of Microarrays in Genomics
Can use microarray to find what alleles (homo and hetero) a person has for a SNP by measuring the level of hybridization to a probe with one allele.  
Can read ChIP data from a microarray.

### Microarray Assessment
1. This code chunk constructs a `SummarizedExperiment` container for expression data on samples taken from different tissues. We'll learn more about this container design later on. For now, just follow the instructions below, adding packages to your installation if any are missing, using `biocLite()`. For example, acquire the `tissuesGeneExpression` package using `biocLite("genomicsclass/tissuesGeneExpression")` if you don't already have it.
```{r, message=FALSE, warning=FALSE}
library(tissuesGeneExpression)
data(tissuesGeneExpression)
library(SummarizedExperiment)
tissSE = SummarizedExperiment(list(rma=e))
colData(tissSE) = DataFrame(tab)
library(hgu133a.db)
sym = mapIds(hgu133a.db, keys=rownames(tissSE), column="SYMBOL", keytype="PROBEID")
nm = mapIds(hgu133a.db, keys=rownames(tissSE), column="GENENAME", keytype="PROBEID")
rowData(tissSE) = DataFrame(symbol=sym, genename=nm) 
```

We can use
```{r}
grep("GAPDH", rowData(tissSE)$symbol, value=TRUE)
```
to count the number of array features that measure expression of gene GAPDH.
How many features in this SummarizedExperiment measure expression of gene H2AFX?
```{r}
hx <- grep("H2AFX", rowData(tissSE)$symbol, value=T)
length(hx)
```

2. Verify that 205436_s_at is the affymetrix code for H2AFX and then consider the following plot. Which of the following relationships are suggested by this plot?
```{r}
par(las=2, mar=c(10,4,2,2))
boxplot(as.numeric(assay(tissSE["205436_s_at",]))~tissSE$Tissue)
```
## Introduction to Next Generation Sequencing Technology
### Next Generation Sequencing Technology 2: Applications of NGS in Genomics
Used for variant detection in a population (SNP discovery).  
RNA-seq is essential NGS of cDNA mapped to a reference.
ChIP-seq using peak-detectors to find hotspots along reference genome.
